[clang] Introduce elementwise ctlz/cttz builtins

frasercrmck opened the pull request:
These builtins are modeled on the clzg/ctzg builtins, which accept an optional second argument. This second argument is returned if the first argument is 0. These builtins unconditionally exhibit zero-is-undef behaviour, regardless of target preference for the other ctz/clz builtins. The builtins have constexpr support.

Fixes #154113

frasercrmck committed:
[clang] Introduce elementwise clz/ctz builtins

These builtins are modeled on the clzg/ctzg builtins, which accept an
optional second argument. This second argument is returned if the first
argument is 0.

tbaederr commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -1484,6 +1484,18 @@ def ElementwiseSubSat : Builtin {
+
+def ElementwiseCtz : Builtin {
+  let Spellings = ["__builtin_elementwise_ctz"];
+  let Attributes = [NoThrow, Const, CustomTypeChecking, Constexpr];
```
There is no constexpr implementation in this patch.

frasercrmck replied:
Oh good catch, thank you. I must have copy/pasted the wrong builtins..

frasercrmck replied:
Ah no, the documentation says that the integer builtins may be called in a constexpr context, so `Constexpr` is correct and the patch needs to account for that. Will update.

frasercrmck replied:
Done - it now supports constexpr evaluation.

frasercrmck commented:
ping, thanks

frasercrmck commented:
Friendly ping

arsenm commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -831,6 +832,14 @@ of different sizes and signs is forbidden in binary and ternary builtins.
+ T __builtin_elementwise_ctz(T x[, T y])        return the number of trailing 0 bits in the first argument. If         integer types
+                                                the first argument is 0 and an optional second argument is provided,
+                                                the second argument is returned. If the first argument is 0 but only
+                                                one argument is provided, the behaviour is undefined.
```
Violations are poison, not instant undefined behavior?

frasercrmck replied:
> Violations are poison, not instant undefined behavior?

Hmm, that's true. I copied this from the documentation for __builtin_clzg/__builtin_ctzg. Is `poison` something that should be mentioned at the language level, though? There's no mention of poison anywhere else in the document.

> Wondering if we should use the names of the llvm intrinsics. If the arguments and behavior are different from the legacy gcc builtins, we don't need to copy the name

That makes sense to me. Alternatively if they were named `builtin_elementwise_c[lt]zg` then it might indicate they behave identically to *those* builtins. But as it stands the use of `ctz` is misleading as it conflates behaviour with the GCC builtins.

Is this RFC territory?

arsenm replied:
On regular __builtin_clz/ctz the gcc docs state "result is undefined". Not the program is undefined

arsenm replied:
Could do an RFC I suppose. I always found the target selection of behavior absurd with the gcc builtins

frasercrmck replied:
> On regular __builtin_clz/ctz the gcc docs state "result is undefined". Not the program is undefined

That makes sense, yep. I've updated to use that language here.

frasercrmck replied:
> Could do an RFC I suppose. I always found the target selection of behavior absurd with the gcc builtins

I've started an RFC: https://discourse.llvm.org/t/rfc-introducing-elementwise-clz-ctz-builtins/85862.

frasercrmck replied:
As expected, nobody has any strong opinions. Neither do I, to be honest. I think either `__builtin_elementwise_ctz` or `__builtin_elementwise_cttz`. Having `g` doesn't make much sense to me if there are no non-g alternatives.

There are already builtins like `__builtin_elementwise_popcount` which map to `llvm.ctpop` so my inclination is to stick with the GCC names. If you feel strongly about having the intrinsic naming I'll go with that.

frasercrmck commented:
ping

efriedma-quic commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -847,6 +848,14 @@ of different sizes and signs is forbidden in binary and ternary builtins.
+ T __builtin_elementwise_clz(T x[, T y])        return the number of leading 0 bits in the first argument. If          integer types
+                                                the first argument is 0 and an optional second argument is provided,
+                                                the second argument is returned. If the first argument is 0 but only
+                                                one argument is provided, the result is undefined.
```
What exactly does "the result is undefined" mean?  The standard has "undefined behavior", or "indeterminate value"(C++)/"indeterminate representation"(C), or in the latest C++ draft "erroneous value".  Nothing defines "the result is undefined", so it's not clear what this means.

arsenm replied:
It should not say it's undefined behavior, the gcc description of the behavior is still "If x is 0, the result is undefined". The resulting call is not marked noundef, it is a poison returning operation. It is not instant undefined behavior, you need to make use of the result to trigger UB

efriedma-quic replied:
Technically speaking, there's some LLVM IR operations which won't trigger UB on a poison value... but it's much narrower than an uninitialized variable.  In particular, poison propagates much more strongly, and it can end up poisoning adjacent fields in memory.  And it's not consistent with any form of "undefined" in any standard.  Because of that, I don't want to document any clang intrinsics as "returns poison".

"the result is undefined" is just that but worse: it's equally messy, and ambiguous.

So I think the only viable alternative to "the behavior is undefined" is "the result is unspecified"... which would mean we have to actually pick a value to return.

efriedma-quic reviewed (approved):
LGTM

arsenm reviewed (commented):
Also does this have the same maybe-defined-maybe-not-depending-on-target-preference as __builtin_clz? Can you document that explicitly if it is or not?

I've always hated that behavior and think it's absurd. My main hesitation is if we're doing something less confusing, could switch to match the llvm intrinsic name

efriedma-quic commented:
> Oh excuse me, there's currently no target-specific behaviour. I missed the HasFallback in

HasFallback only applies to the two-argument form.

------------

I don't have a strong opinion about the names.

The whole "fallback behavior" thing was written back in the early days of clang, before we even had ubsan.  I doubt we would handle that the same way today... and I don't think we need to carry it forward to new variants.  If the user cares, they can explicitly request `__builtin_elementwise_ctz(x, 32)` or whatever.

frasercrmck commented:
> > Oh excuse me, there's currently no target-specific behaviour. I missed the HasFallback in
> 
> HasFallback only applies to the two-argument form.

Oh yes, thank you. I've fixed that now. The behaviour is unconditionally zero-is-undef.
> 
> I don't have a strong opinion about the names.
> 
> The whole "fallback behavior" thing was written back in the early days of clang, before we even had ubsan. I doubt we would handle that the same way today... and I don't think we need to carry it forward to new variants. If the user cares, they can explicitly request `__builtin_elementwise_ctz(x, 32)` or whatever.

The builtins are now renamed to cttz/ctlz.

@arsenm I haven't at this point updated the documentation. I note that it always says "It is undefined behaviour if the first argument is 0 and no second argument is provided.". My literal reading of this is already that it's always undefined behaviour for all targets (as you're requesting), but perhaps with the baggage of the old builtins in mind it's still not clear? What do you think?

arsenm commented:
> My literal reading of this is already that it's always undefined behaviour for all targets (as you're requesting),

That's not what I'm requesting, it would be an undefined *value*. It is not instant undefined behavior, you would get UB on use of that value

frasercrmck commented on clang/test/Sema/constant-builtins-vector.cpp:
```diff
@@ -860,3 +860,41 @@ static_assert(__builtin_elementwise_sub_sat(0U, 1U) == 0U);
+static_assert(__builtin_elementwise_ctlz((char)128) == 0x0);
+static_assert(__builtin_bit_cast(unsigned, __builtin_elementwise_ctlz((vector4char){1, 4, 127, (char)128})) == (LITTLE_END ? 0x00010507 : 0x07050100));
+
+constexpr int clz0 = __builtin_elementwise_ctlz(0);
```
Good idea, I've added those tests now.

arsenm commented:
I'd view assertively stating its undefined behavior as a move away from the status quo.  For `__builtin_clz`, GCC states `the result is undefined`. Clang appears to not have separate documentation for that exact spelling. It does have documentation for `__builtin_clzg`, which has the undefined behavior phrasing. 

> We don't currently have a definition of what it means to "use" a value in a way that would cause undefined behavior with poison.

Which means stating it produces undefined behavior less accurate?

> If you have some plan to use this notion of "undefined value" for something other than ctlz/cttz,

There are a handful of other intrinsics which are documented as returning `freeze poison`

RKSimon reviewed (commented):
@frasercrmck please can you update this patch against trunk

frasercrmck commented:
> @frasercrmck please can you update this patch against trunk

done, thanks for noticing.

efriedma-quic commented:
I don't know of any such native instructions, no, but some expansions are slightly less efficient if zero inputs are allowed.  See llvm/test/CodeGen/RISCV/rvv/cttz-sdnode.ll .

arsenm reviewed (approved):
I think the only issue is documentation phrasing, but the existing builtins already have internally inconsistent phrasing, so that can be addressed later

jhuber6 commented:
Dropping in after the fact, is there a reason we called this `__builtin_elementwise_ctlz` instead of `__builtin_elementwise_clzg`? The builtin is just `clzg` done on each element so the name is confusing me.

jhuber6 commented:
> > Dropping in after the fact, is there a reason we called this `__builtin_elementwise_ctlz` instead of `__builtin_elementwise_clzg`? The builtin is just `clzg` done on each element so the name is confusing me.
> 
> It matches the llvm intrinsic name, and the second argument is a difference from clz

That is what `clzg` does, unless I'm missing something https://godbolt.org/z/x1fqTxc43.
